src/main/java/com/google/devtools/build/lib/analysis/constraints/RuleContextConstraintSemantics.java

```diff
@@ -808,4 +825,147 @@ private static void addSelectValuesToSet(BuildType.Selector<?> select, final Set
+      throws ActionConflictException {
+    if ("toolchain".equals(ruleContext.getRule().getRuleClass())) {
+      return null;
+    }
+
+    // This is incompatible if explicitly specified to be.
+    if (ruleContext.attributes().has("target_compatible_with")) {
+      Iterable<ConstraintValueInfo> constraintValues =
+          PlatformProviderUtils.constraintValues(
+              ruleContext.getPrerequisites("target_compatible_with"));
+      for (ConstraintValueInfo constraintValue : constraintValues) {
+        if (!ruleContext.targetPlatformHasConstraint(constraintValue)) {
+          return createIncompatibleConfiguredTarget(ruleContext, null, constraintValue);
+        }
+      }
+    }
+
+    // This is incompatible if one of the dependencies is as well.
+    for (ConfiguredTargetAndData infoCollection : prerequisiteMap.values()) {
+      ConfiguredTarget dependency = infoCollection.getConfiguredTarget();
+      if (dependency instanceof OutputFileConfiguredTarget) {
+        // For generated files, we want to query the generating rule for providers. genrule() for
+        // example doesn't attach providers like IncompatiblePlatformProvider to its outputs.
+        dependency = ((OutputFileConfiguredTarget) dependency).getGeneratingRule();
+      }
+      if (dependency.getProvider(IncompatiblePlatformProvider.class) != null) {
+        return createIncompatibleConfiguredTarget(ruleContext, dependency, null);
+      }
+    }
+
+    return null;
+  }
+
+  /**
+   * A helper function for incompatibleConfiguredTarget() to actually create the incompatible {@link
+   * ConfiguredTarget}.
+   *
+   * @param ruleContext analysis context for the rule
+   * @param targetResponsibleForIncompatibility the target that is responsible this target's
+   *     incompatibility. If null, that means that target is responsible for its own
+   *     incompatibility. I.e. it has constraints in target_compatible_with that were not satisfied
+   *     on the target platform. This must be null if violatedConstraint is set. This must be set if
+   *     violatedConstraint is null.
+   * @param violatedConstraint the constraint that the target platform doesn't satisfy. This must be
+   *     null if targetResponsibleForIncompatibility is set.
+   * @throws ActionConflictException if the underlying {@link RuleConfiguredTargetBuilder}
+   *     encounters a problem when assembling a dummy action for the incompatible {@link
+   *     ConfiguredTarget}.
+   */
+  private static ConfiguredTarget createIncompatibleConfiguredTarget(
+      RuleContext ruleContext,
+      ConfiguredTarget targetResponsibleForIncompatibility,
+      ConstraintValueInfo violatedConstraint)
+      throws ActionConflictException {
+    // Create a dummy ConfiguredTarget that has the IncompatiblePlatformProvider set.
+    ImmutableList<Artifact> outputArtifacts = ruleContext.getOutputArtifacts();
+
+    if (ruleContext.isTestTarget() && outputArtifacts.size() == 0) {
```

Comment:
Nit: use outputArtifacts.isEmpty() instead of checking the size.

Reply:
Done